File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.java
import java.util.*;

class Solution {
    public int[] solution(int[] fees, String[] records) {
        Map<String, Integer> inTimeMap = new HashMap<>();
        Map<String, Integer> totalTimeMap = new HashMap<>();
        
        // 기록 처리
        for (String record : records) {
            String[] carInfo = record.split(" ");
            String carNum = carInfo[1];
            int hour = Integer.parseInt(carInfo[0].substring(0, 2));
            int min = Integer.parseInt(carInfo[0].substring(3, 5));
            int time = hour * 60 + min;
            String state = carInfo[2];
            
            if (state.equals("IN")) {
                inTimeMap.put(carNum, time);
            } else {
                int inTime = inTimeMap.remove(carNum);
                totalTimeMap.put(carNum, totalTimeMap.getOrDefault(carNum, 0) + (time - inTime));
            }
        }
        
        // 출차 기록이 없는 차량 처리
        for (String carNum : inTimeMap.keySet()) {
            int inTime = inTimeMap.get(carNum);
            int endTime = 23 * 60 + 59;
            totalTimeMap.put(carNum, totalTimeMap.getOrDefault(carNum, 0) + (endTime - inTime));
        }
        
        // 요금 계산
        Map<String, Integer> feeMap = new TreeMap<>();
        for (String carNum : totalTimeMap.keySet()) {
            feeMap.put(carNum, calculateFee(fees, totalTimeMap.get(carNum)));
        }
        
        // 결과 배열 생성
        int[] answer = new int[feeMap.size()];
        int idx = 0;
        for (int fee : feeMap.values()) {
            answer[idx++] = fee;
        }
        
        return answer;
    }
    
    private int calculateFee(int[] fees, int time) {
        int initTime = fees[0];
        int initFee = fees[1];
        int unitTime = fees[2];
        int unitFee = fees[3];
        
        if (time <= initTime) {
            return initFee;
        } else {
            return initFee + (int) Math.ceil((time - initTime) / (double) unitTime) * unitFee;
        }
    }
}

File: 프로그래머스/2/92341. 주차 요금 계산/주차 요금 계산.py
import math

def solution(fees, records):
    answer = []
    # car_time: 차량별 누적 주차 시간(분)
    car_time = {}
    # current_car_time: 입장(IN) 시간(분)을 저장. OUT이면 삭제
    current_car_time = {}
    
    for record in records:
        time_str, car_no, status = record.split()
        t = minute_converter(time_str)
        
        if status == 'IN':
            # 입장: 입장 시간만 저장
            current_car_time[car_no] = t
            # 누적시간 초기화(처음 보는 차량이면 0으로 초기화)
            if car_no not in car_time:
                car_time[car_no] = 0
        else:  # OUT
            # 안전하게 IN이 있었는지 확인
            in_time = current_car_time.get(car_no)
            if in_time is None:
                # 비정상 데이터지만 무시하거나 continue
                continue
            duration = t - in_time
            car_time[car_no] = car_time.get(car_no, 0) + duration
            # 현재 입장 기록 제거
            del current_car_time[car_no]
    
    # 23:59까지 나가지 않은 차량 처리
    end_of_day = minute_converter("23:59")
    for car_no, in_time in list(current_car_time.items()):
        duration = end_of_day - in_time
        car_time[car_no] = car_time.get(car_no, 0) + duration
        # del current_car_time[car_no]  # optional
    
    # 요금 계산
    default_minute, default_fee, unit_time, unit_fee = fees
    # 차량번호 오름차순 정렬
    for car_no in sorted(car_time.keys()):
        used_minute = car_time[car_no]
        if used_minute <= default_minute:
            answer.append(default_fee)
        else:
            extra = used_minute - default_minute
            fee_units = math.ceil(extra / unit_time)
            total_fee = default_fee + fee_units * unit_fee
            answer.append(total_fee)
    
    return answer

def minute_converter(time):
    h, m = time.split(":")
    return int(h) * 60 + int(m)
